TC-SU-4.1 python test

juandediosg opened the pull request:
### Summary

This PR is focused on implementing and validating the TC-SU-4.1
Fixes: https://github.com/project-chip/matter-test-scripts/issues/659
Test plan: [3.10.1. [TC-SU-4.1] Verifying Cluster Attributes on OTA-R(DUT)](https://github.com/CHIP-Specifications/chip-test-plans/blob/master/src/softwareupdate.adoc#3101-tc-su-41-verifying-cluster-attributes-on-ota-rdut)
Test plan updates: https://github.com/CHIP-Specifications/chip-test-plans/pull/5807

### Scope

### Related Issues & PRs
**Umbrealla:** [#648](https://github.com/project-chip/matter-test-scripts/issues/648) TC-SU and TC-BDX: Automate and re-think (umbrella) 
**yaml:** https://github.com/project-chip/connectedhomeip/blob/master/src/app/tests/suites/certification/Test_TC_SU_4_1.yaml
**matter OTA:** https://github.com/project-chip/connectedhomeip/blob/e560a107149315e2b8da09dc8e53afc3deba5a4d/docs/platforms/esp32/ota.md
**Bug:** [[BUG] WriteAttribute updates DefaultOTAProviders attribute despite ConstraintError in DefaultOTAProviders cluster #40294](https://github.com/project-chip/connectedhomeip/issues/40294) (CLOSED)


#### Testing
To run the automated TC-SU-4.1 test, follow these steps:

### Reproduce Test
To reproduce the issue, please follow the prerequisites and test steps described below.

#### Prerequisite: Build and run Matter Project
```bash
# Setup your Python environment
# Run Bootstrap
source scripts/bootstrap.sh
# Activate
source scripts/activate.sh
# Env
source out/python_env/bin/activate
```

#### Prerequisite: OTA

**Build  Components:**
```bash
# Build OTA Provider
scripts/examples/gn_build_example.sh examples/ota-provider-app/linux out/debug chip_config_network_layer_ble=false

# Build OTA Requestor
scripts/examples/gn_build_example.sh examples/ota-requestor-app/linux out/debug chip_config_network_layer_ble=false

# Build Chip Tool
scripts/examples/gn_build_example.sh examples/chip-tool out/chiptool 'chip_mdns="platform"'
  ```
**Create OTA Images and/or Required OTA Images**
```bash
N/A Only is required OTA_REQUESTOR_APP
```

#### Run Test

```bash
# Launch OTA Requestor from Terminal 1:
./out/debug/chip-ota-requestor-app --discriminator 1234 --passcode 20202021 --KVS /tmp/chip_kvs_requestor

# Launch Python test from Terminal 2:
python3 src/python_testing/TC_SU_4_1.py \
  --commissioning-method on-network \
  --discriminator 1234 \
  --passcode 20202021
```

juandediosg committed:
Initial commit..

juandediosg committed:
Initial commit with partial implementation.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,37 @@
+#       --trace-to perfetto:${TRACE_TEST_PERFETTO}.perfetto
+#     factory-reset: true
+#     quiet: true
+# === END CI TEST ARGUMENTS ===
```
Implementation Updated but not functional yet

juandediosg committed:
Implemented Step 1, still not functional.

juandediosg committed:
Implemented Steps from 2 to 4, partial functional.

juandediosg committed:
Refactor: Implemented Steps from 1 to 2 functional, with TH3 added and commissioning.

juandediosg committed:
Refactor: Implemented Steps from 1 to 4 working as expected.

cecille commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,289 @@
+        # ------------------------------------------------------------------------------------
+        # Manual Setup
+        # ------------------------------------------------------------------------------------
+        # 1. Launch OTA Provider (TH2) from Terminal 1:
```
I don't think you actually need any of this - my suggestion would be to remove step 9 and combine it with another test where downloading the image is required. It would be much better to consolidate the transfer steps as much as possible since transferring and applying an image is fairly expensive. This would then mean that the providers here can be imaginary - just some random node id that we're writing and reading back.

juandediosg committed:
Implemented Step 5.

juandediosg committed:
Refactor: Implemented Step 5.

juandediosg committed:
Refactor: Implemented partial Step 6.

juandediosg committed:
Refactor: Implemented partial Step 4 debbuging for potential bug.

juandediosg committed:
Debug: Investigate potential bug in step 4 impacting step 6.

juandediosg committed:
Debug: Investigate potential bug in step 5 impacting step 6.

juandediosg committed:
Refactor: Updated Step 4 based on bug 40294.

juandediosg committed:
Updated Step 6

juandediosg committed:
Updated Step 7

juandediosg committed:
Updated Step 8

juandediosg committed:
Step 9 skipped per test plan refinement.

juandediosg committed:
Fix: Restyle errors.

juandediosg committed:
Fix: Restyle errors v2.

juandediosg committed:
Fix: code-lints errors.

juandediosg committed:
Fix: Restyle errors v3.

juandediosg committed:
Fix: code-lints errors v2.

juandediosg committed:
Fix: Corrected CI test arguments to use endpoint 0.

juandediosg committed:
Merge branch 'master' into TC_SU_4_1_PythonTest

juandediosg committed:
Fix: Updated CI test arguments.

r-recchia commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,483 @@
+    """
+
+    """
+    cluster_otap = Clusters.OtaSoftwareUpdateProvider
```
Unused variable `cluster_otap`

juandediosg replied:
Fixed: Keeping `cluster_otap` as a commented line for reference in OTA test.

r-recchia commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,483 @@
+
+    """
+    cluster_otap = Clusters.OtaSoftwareUpdateProvider
+    cluster_otar = Clusters.OtaSoftwareUpdateRequestor
```
Is this the right place? `cluster_otar` is only used in `test_TC_SU_4_1` method

r-recchia commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,483 @@
+    cluster_otap = Clusters.OtaSoftwareUpdateProvider
+    cluster_otar = Clusters.OtaSoftwareUpdateRequestor
+
+    async def write_acl(self, controller, acl):
+        """
+        Writes the Access Control List (ACL) to the DUT device using the specified controller.
+
+        Args:
+            controller: The Matter controller (e.g., th1, th4) that will perform the write operation.
+            acl (list): List of AccessControlEntryStruct objects defining the ACL permissions to write.
+
+        Raises:
+            AssertionError: If writing the ACL attribute fails (status is not Status.Success).
+        """
+        result = await controller.WriteAttribute(
+            self.dut_node_id,
+            [(0, Clusters.AccessControl.Attributes.Acl(acl))]
+        )
+        asserts.assert_equal(result[0].Status, Status.Success, "ACL write failed")
```
This function is writing and validating but the name only contains write. Consider changing the function name or moving the assertion outside this function

juandediosg replied:
Fixed: Function has been removed as we are now using `create_acl_entry` from the SUTestBase class.

r-recchia commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,483 @@
+        logger.info(f'Step #2 - DefaultOTAProviders value read from DUT (TH1): {actual_otap_info}')
+
+        # Verify that the provider list is not empty (expecting at least one provider)
+        asserts.assert_true(len(actual_otap_info) > 0, "DefaultOTAProviders list is empty")
```
Maybe use assertions like matter_asserts.assert_list(list, description: str, min_length: 1) or assert_greater_equal

juandediosg replied:
I prefer to keep `assert_true(len(...) > 0)` because it is clear and explicit with the comment. Also, Mobly doesn’t have an `assert_list` method.

r-recchia commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,483 @@
+        acl = [admin_acl, view_acl]
+
+        # Write ACL using TH1 since it has permission; this grants TH4 the needed access.
+        resp = await self.write_acl(th1, acl)
```
`write_acl` function is used just one time. Consider using it in more places if needed or removing the function if it is just one line

juandediosg replied:
I prefer to keep the `write_acl` function even if it’s currently used only once because it clearly encapsulates a necessary setup step outside the main test logic, essential for the proper execution of the test. Besides, according to PEP 8, breaking code into small, focused functions improves readability and maintainability, . Also, it can serve as a clear reference for future tests or similar setups.

juandediosg replied:
Fixed: Function has been removed as we are now using `create_acl_entry` from the SUTestBase class.

r-recchia commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,483 @@
+        logger.info(f'Step #1 - Read actaul DefaultOTAProviders value on DUT (TH1): {actual_otap_info}')
+
+        # TH2 is the OTA Provider (NodeID=1) for fabric 1
+        provider_th2_for_fabric1 = Clusters.OtaSoftwareUpdateRequestor.Structs.ProviderLocation(
```
Consider using cluster_otar

juandediosg replied:
Fixed:  Updated to use `self.cluster_otar`

r-recchia commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,483 @@
+        )
+
+        # DefaultOTAProviders attribute with the provider list
+        attr = Clusters.OtaSoftwareUpdateRequestor.Attributes.DefaultOTAProviders(value=[provider_th2_for_fabric1])
```
Same here and in other parts of the code

juandediosg replied:
Fixed: Updated to use self.cluster_otar

r-recchia commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,483 @@
+        logger.info(f'Step #7 - Read UpdatePossible attribute on DUT using TH3 (fabric 2): {update_possible_th3}')
+
+        # Verify UpdatePossible is true
+        asserts.assert_equal(update_possible_th3, True, "Expected UpdatePossible to be True on fabric 2")
```
Use assert_true

juandediosg replied:
Fix: Replaced assert_equal with assert_true

r-recchia commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,483 @@
+        attr = Clusters.OtaSoftwareUpdateRequestor.Attributes.DefaultOTAProviders(value=[provider_th3_for_fabric2])
+
+        # Write updated DefaultOTAProviders attribute to DUT (TH1)
+        resp = await th3.WriteAttribute(
+            attributes=[(0, attr)],
+            nodeid=self.dut_node_id,
+        )
```
It would be nice to use `write_single_attribute` and add an extra parameter to select the `default_controller` if needed. Maybe in a follow up issue

juandediosg committed:
docs: Added docstring.

juandediosg committed:
docs: add comment explaining unused OTA cluster variable kept for future reference.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,483 @@
+        self.step(6)
+
+        # Create Empty Providers list
+        providers_list_empty = []
+        logger.info(f'Step #6 - Providers list updated with provider empty: {providers_list_empty}')
```
The log is intentional to indicate the state of `providers_list_empty` and aid debugging in the step.

juandediosg committed:
Fix: Use self.cluster_otar instead of direct reference.

juandediosg committed:
Fix: Replace assert_equal with assert_true.

juandediosg reviewed (commented):
I have addressed all code review comments and made the necessary updates.

juandediosg committed:
Fix: Updated CI arguments for app.

juandediosg committed:
CI: Added OTA_REQUESTOR and OTA_PROVIDER to 'Generate an argument environment file' workflow to file 'test_env.yaml'.

juandediosg committed:
Merge branch 'master' into TC_SU_4_1_PythonTest

jtrejoespinoza-grid commented on .github/workflows/tests.yaml:
```diff
@@ -706,6 +706,8 @@ jobs:
                   echo "CLOSURE_APP: objdir-clone/linux-x64-closure-${BUILD_VARIANT}-tsan-clang-test/closure-app" >> /tmp/test_env.yaml
                   echo "PUSH_AV_SERVER: objdir-clone/push_av_server/server.py" >> /tmp/test_env.yaml
+                  echo "OTA_REQUESTOR: objdir-clone/linux-x64-ota-requestor-${BUILD_VARIANT}-tsan-clang-test/chip-ota-requestor-app" >> /tmp/test_env.yaml
+                  echo "OTA_PROVIDER: objdir-clone/linux-x64-ota-provider-${BUILD_VARIANT}-tsan-clang-test/chip-ota-provider-app" >> /tmp/test_env.yaml
```
I see this step to add the path for the `OTA_REQUESTOR` and the `OTA_PROVIDER` but I do not see the step when the apps are built at `- name: Build Apps`. Please add the step to build the apps to be available for the Tests.

juandediosg replied:
I see the build step for the OTA apps at lines 247-248 in this file (_tests.yaml_), which are already built in the `Build Apps` step. Here is the relevant snippet:

```yaml
- name: Build Apps
  run: |
    ./scripts/run_in_build_env.sh \
      "./scripts/build/build_examples.py \
        --target linux-x64-ota-provider-${BUILD_VARIANT} \
        --target linux-x64-ota-requestor-${BUILD_VARIANT} \
        --copy-artifacts-to objdir-clone \
      "
```
This step builds the binaries that are later used in the `Generate an argument environment` file step via the `OTA_REQUESTOR` and `OTA_PROVIDER` variables.

Once the CI jobs run, I will confirm if everything works as expected or if any adjustments are needed.

juandediosg replied:
Fixed: the build were implemented on [Adds OTA_REQUESTOR and OTA_PROVIDER on CI build
#672](https://github.com/project-chip/matter-test-scripts/issues/672)

juandediosg reviewed (commented):
Clarified with @jtrejoespinoza-grid  that the OTA `build` step is already part of the workflow.

juandediosg committed:
CI: Added build step for chip-ota-requestor-app.

juandediosg committed:
Fix: Refactor build step for linux-x64-ota-requestor to enable test run.

asirko-soft commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,492 @@
+        actual_otap_info = await self.read_single_attribute_check_success(
+            cluster=self.cluster_otar,
+            attribute=self.cluster_otar.Attributes.DefaultOTAProviders)
+        logger.info(f'Step #1 - Read actaul DefaultOTAProviders value on DUT (TH1): {actual_otap_info}')
```
Please fix the typo "actaul" -> "actual"

juandediosg replied:
Fixed: Comment has been updated.

asirko-soft commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,492 @@
+
+        # Verify DefaultOTAProviders attribute on the DUT after write (TH3 on Fabric 2)
+        update_possible_th3 = await self.read_single_attribute_check_success(
+            dev_ctrl=th4,
```
Please double check this place, copilot is right, the variable name is update_possible_th3, comment suggests intention to check TH3 too, but TH4 is checked: dev_ctrl=th4. These should be aligned.

juandediosg replied:
Clarification: Comment has been updated to clarify that although the test case mentions TH3 for Fabric 2, we are using the existing TH4 session to read the attribute. This avoids opening an additional session and the verification remains valid for Fabric 2.

asirko-soft commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,492 @@
+        logger.info(f'Step #7 - Read UpdatePossible attribute on DUT using TH3 (fabric 2): {update_possible_th3}')
+
+        # Verify UpdatePossible is true
+        asserts.assert_true(update_possible_th3, "Expected UpdatePossible to be True on fabric 2")
```
Please check this copilot detection as well, something is not right there.

juandediosg replied:
Same as above comment, this is intentional. we use TH4 session to read `Fabric 2` attributes to avoid opening another session; the verification for `Fabric 2` remains valid. Comments updated for clarity.

asirko-soft reviewed (changes requested):
Please fix typos and review the copilot detections. Other than that, looks good to me

juandediosg committed:
Clean branch: revert accidental merges; restore branch to correct state.

juandediosg committed:
Merge branch 'master' into TC_SU_4_1_PythonTest

juandediosg committed:
Fix: Address Ruff.

juandediosg committed:
Fix: Updated test to be compatible with latest master.

juandediosg committed:
Merge branch 'master' into TC_SU_4_1_PythonTest

juandediosg committed:
Refactor: In Step 5 reuse create_acl_entry for TH4 with Admin + View permissions.

juandediosg committed:
Refactor:
- Updated CI test arguments to remove --nodeId, vendor and product ID, and port.
- Removed unused imports and the write_acl() function; now using acl_entry for ACLs.
- Reordered and restyled library calls for consistent style.

juandediosg committed:
Refactor and test improvements:
- Refactor steps 1, 3, and 5 to use set_default_ota_providers_list fn and improve logging.
- Add explanatory note in step 6 regarding multiple providers and ConstraintError.
- Update teardown_test to clear KVS after test.
- Include OTA_REQUESTOR_APP environment variable in CI.

juandediosg committed:
Update: Added clarifications in Step 7 feedback.

juandediosg reviewed (commented):
Refactored the test to use `SUTestBase`. All previous comments have been addressed. Added clarifications in some feedback. Copilot noise comments still pending cleanup.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,468 @@
+            cluster=self.cluster_otar,
+            attribute=self.cluster_otar.Attributes.DefaultOTAProviders)
+
+        logger.info(f'Step #6 - Read DefaultOTAProviders attribute on DUT using TH4: {th4_actual_otap_info}')
```
N/A, Current logic is intentional. Step 6 writes an empty provider list to Fabric 2 (via TH3) and reads Fabric 1 (via TH4) to verify the original provider is intact. Additional assertions are not required per the test case requirements.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,455 @@
+    async def teardown_test(self):
+        # Clear provider KVS after test to avoid interference with subsequent tests
+        self.clear_kvs()
+        await super().teardown_test()
```
N/A. No fabrics are created by this test; all fabrics are pre-existing and managed by the harness.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,455 @@
+            TestStep(6, "TH sends a write request for the DefaultOTAProviders Attribute with an empty provider list on the second fabric to the DUT. "
+                     "TH sends a read request to read the DefaultOTAProviders Attribute on the first and second fabric to the DUT.",
+                     "Verify that the write operation for the attribute works and DUT does not respond with any errors."
+                     "Verify that the attribute value is set to TH4 as the default OTA provider for the first fabric and none for the second fabric."),
```
N/A, acknowledged. Current step is intentional from test plan.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,455 @@
+
+        # Verify UpdatePossible is true
+        asserts.assert_true(update_possible_th3, "Expected UpdatePossible to be True on fabric 2")
+
```
Removed redundant UpdatePossible read. The test plan only requires a single verification and we using TH4

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,457 @@
+
+        self.step(5)
+
+        # Establishing TH4 controller - TH4, NodeID=4, Fabric=2
```
Updated comment to correctly reflect that `TH4` is on `Fabric 1`.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,457 @@
+            TestStep(3, "TH sends a write request for the DefaultOTAProviders Attribute on the second fabric to the DUT. TH3 is set as the default Provider for the fabric.",
+                     "Verify that the write operation for the attribute works and DUT does not respond with any errors."),
+            TestStep(4, "TH sends a read request to read the DefaultOTAProviders Attribute on the first and second fabric to the DUT."
+                     "Verify that the attribute value is set to TH2 as the default OTA provider for the first fabric and TH3 for the second fabric."),
```
Step comment has been updated.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,473 @@
+                     "TH4 is the first Provider location and TH2 is the second Provider location in the same write request on the first fabric. "
+                     "TH sends a read request to read the DefaultOTAProviders Attribute on the first and second fabric to the DUT.",
+                     "Verify that the write operation fails with CONSTRAINT_ERROR status code 0x87.\n"
+                     "Verify that the attribute value is set to TH3 as the default OTA provider for the second fabric and either of TH2 or TH4 for the first fabric."),
```
N/A, acknowledged. Current step definition is intentional based on test plan.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,468 @@
+            TestStep(6, "TH sends a write request for the DefaultOTAProviders Attribute with an empty provider list on the second fabric to the DUT. "
+                     "TH sends a read request to read the DefaultOTAProviders Attribute on the first and second fabric to the DUT.",
+                     "Verify that the write operation for the attribute works and DUT does not respond with any errors."
+                     "Verify that the attribute value is set to TH4 as the default OTA provider for the first fabric and none for the second fabric."),
```
N/A, acknowledged. Current step definition is intentional based on test plan.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,468 @@
+
+        self.step(5)
+
+        # Establishing TH4 controller - TH4, NodeID=4, Fabric=2
```
Updated comment to correctly reflect that TH4 is on Fabric 1.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,473 @@
+                     "Verify that the attribute value is set to TH2 as the default OTA provider for the fabric."),
+            TestStep(3, "TH sends a write request for the DefaultOTAProviders Attribute on the second fabric to the DUT. TH3 is set as the default Provider for the fabric.",
+                     "Verify that the write operation for the attribute works and DUT does not respond with any errors."),
+            TestStep(4, "TH sends a read request to read the DefaultOTAProviders Attribute on the first and second fabric to the DUT."
```
N/A, acknowledged. Current step format is intentional based on test plan.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,468 @@
+
+        # Verify DefaultOTAProviders attribute on the DUT after write (TH3 on Fabric 2)
+        update_possible_th3 = await self.read_single_attribute_check_success(
+            dev_ctrl=th4,
```
Removed redundant `UpdatePossible` read. The test plan only requires a single verification and we using `TH4`.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,492 @@
+        logger.info(f'Step #7 - Read UpdatePossible attribute on DUT using TH3 (fabric 2): {update_possible_th3}')
+
+        # Verify UpdatePossible is true
+        asserts.assert_true(update_possible_th3, "Expected UpdatePossible to be True on fabric 2")
```
Removed redundant UpdatePossible read. The test plan only requires a single verification and we using TH4.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,468 @@
+
+        self.step(8)
+        # Step #8 - Read UpdateState from TH4 (fabric 1)
+        t4_update_state = await self.read_single_attribute_check_success(
```
Renamed variable to `th4_update_state` for consistency with existing `TH4` naming convention in the test.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,468 @@
+        actual_otap_info = await self.read_single_attribute_check_success(
+            cluster=self.cluster_otar,
+            attribute=self.cluster_otar.Attributes.DefaultOTAProviders)
+        logger.info(f'Step #1 - Read actaul DefaultOTAProviders value on DUT (TH1): {actual_otap_info}')
```
Logger comment has been updated.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,473 @@
+
+        self.step(1)
+
+        # Establishing TH1 controller - DUT is TH1, NodeID=2, Fabric=1
```
Comment has been updated.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,473 @@
+            attribute=self.cluster_otar.Attributes.DefaultOTAProviders)
+
+        logger.info(f'Step #6 - Read DefaultOTAProviders attribute on DUT using TH4: {th4_actual_otap_info}')
+
```
N/A. Step 6 aligns with the test plan. Validation of the provider configuration impact is covered in Step 7.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,473 @@
+
+        # Write ACL using TH1 since it has permission; this grants TH4 the needed access.
+        resp = await self.write_acl(th1, acl)
+        logger.info(f'Step #5 - TH4 have the necessary permissions to perform operation: {resp}')
```
`write_acl` is not longer used because the test now uses the SUTestBase class `create_acl_entry`

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,457 @@
+        # Verify that the provider list is empty
+        asserts.assert_true(len(actual_provider) == 0, "DefaultOTAProviders list is not empty")
+
+        # Verify the actual provider matches the expected OTA Provider (TH3)
```
Comment has been updated.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,457 @@
+        # The rest of the list is ignored, and the original value remains unchanged.
+
+        # Verify DefaultOTAProviders attribute on the DUT after write (TH4 on Fabric 1) in this case should be the original from TH2 fabric 1
+        # Read fabric 1 with TH4 and verify that the provider is TH1 (providerNodeID=1 , fabricIndex=1 )
```
Comment has been updated.

juandediosg committed:
Update: Clean up test comments and logging,and address Copilot suggestions.

juandediosg committed:
Update: Code-lints and restyled.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,434 @@
+            provider_node_id=self.dut_node_id,
+            acl_entries=acl_list
+        )
+        logger.info(f'Step #5 - TH4 has Admin + View permissions: {resp}')
```
N/A, acknowledged. Current logic is intentional.
Assert is covered in the `create_acl_entry` fn.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,434 @@
+            provider_node_id=self.dut_node_id,
+            acl_entries=acl_list
+        )
+        logger.info(f'Step #5 - TH4 has Admin + View permissions: {resp}')
```
N/A, acknowledged. Current comment is intentional.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,434 @@
+            cluster=self.cluster_otar,
+            attribute=self.cluster_otar.Attributes.UpdatePossible)
+
+        logger.info(f'Step #7 - Read UpdatePossible attribute on DUT using TH4 (fabric 1): {update_possible_th4}')
```
N/A, acknowledged. Current logic is intentional.
only log `UpdatePossible`, assertion is not required for this step per current test plan.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,434 @@
+            attribute=self.cluster_otar.Attributes.DefaultOTAProviders)
+
+        logger.info(f'Step #6 - Read DefaultOTAProviders attribute on DUT using TH4: {th4_actual_otap_info}')
+
```
N/A, acknowledged.  Current logic is intentional based on test plan. no additional assert is needed.

juandediosg committed:
Update: Addressed and clarified Copilot comments, no functional changes to test logic.

khodya commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,432 @@
+        # Write DefaultOTAProviders to DUT using base class helper function
+        await self.set_default_ota_providers_list(
+            controller=th1,
+            provider_node_id=1,   # TH2 is the OTA Provider (NodeID=1)
```
nit: `1` can be a variable (`th2_node_id`) -- should it be `dut_note_id`?

khodya commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,432 @@
+        # Establishing TH3 controller - TH3, NodeID=3, Fabric=2
+        th3_certificate_authority = self.certificate_authority_manager.NewCertificateAuthority()
+        th3_fabric_admin = th3_certificate_authority.NewFabricAdmin(vendorId=0xFFF1, fabricId=th1.fabricId + 1)
+        th3 = th3_fabric_admin.NewController(nodeId=3, useTestCommissioner=True)
```
should it be `nodeId=th2_node_id + 1` ?

khodya commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,432 @@
+        # Reference object for verifying DefaultOTAProviders (TH3, NodeId 3,  Fabric 2)
+        provider_th3_for_fabric2 = self.cluster_otar.Structs.ProviderLocation(
+            providerNodeID=th3.nodeId,
+            endpoint=0,
```
this test will run on endpoint 0 only. is this expected?

juandediosg replied:
Yes, endpoint 0 is expected. OTA clusters run on this endpoint by default.

khodya commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,432 @@
+        # Write the Access Control List (ACL) to the DUT device for TH4 controller
+        # using the SUTestBase function create_acl_entry, passing both Admin and View permissions
+        # so that TH4 can perform necessary operations.
+        acl_list = [
```
why do you need to write the `acl_list`? is this in the test plan?

juandediosg replied:
TH4 is a new node in Fabric 1 but does not participate in commissioning, so it does not receive permissions automatically. Without the ACL, TH4 would not be authorized to read/write the OTA Software Update Requestor attributes. For that reason we add a new ACL to grant the permissions to node=4.

The test plan mentioned in the "Test setup" section https://github.com/CHIP-Specifications/chip-test-plans/blob/master/src/softwareupdate.adoc#test-setup-13

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,432 @@
+
+        # Establishing TH4 controller - TH4, NodeID=4, Fabric=1
+        fabric_admin = self.default_controller.fabricAdmin
+        th4 = fabric_admin.NewController(nodeId=4)
```
I will refactor the NodeIDs to use variables for clarity and maintainability:
```
th2_node_id = 1                # nodeId=1
th3_node_id = th1_node_id + 1  # nodeId=3
th4_node_id = th3_node_id + 1  # nodeId=4
```
This avoids hardcoded numbers and makes the relationships between nodes more explicit.

juandediosg reviewed (commented):
@khodya 
I have addressed your feedback and added some clarifications in a few comments.
I'm going to work on the following refactorings:

- Refactor the Provider NodeIDs to use variables `th2_node_id`, `th3_node_id`, and `th4_node_id` for clarity and maintainability.
- Extract the `FabricID` to a variable `th2_fabric_id`.

Feel free to take a look and share a new round of feedback.

juandediosg committed:
Refactor based on feedback:
- Refactor the Provider NodeIDs to use variables th2_node_id, th3_node_id, and th4_node_id for clarity and maintainability..
- Extract the FabricID for TH3 to a variable th2_fabric_id.

juandediosg committed:
Update: Removed comments for clean-up.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,434 @@
+        th4 = fabric_admin.NewController(nodeId=th4_node_id)
+
+        th4_fabric_id = th4.fabricId
+        logger.info(f'Step #5 - TH4 NodeID: {th4.nodeId}')
```
N/A, acknowledged. Current logic is intentional.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,434 @@
+        logger.info(f'Step #6 - Read DefaultOTAProviders attribute on DUT using TH4: {th4_actual_otap_info}')
+
+        self.step(7)
+        # Step #7: Verify UpdatePossible attribute.
```
N/A, acknowledged. Current logic is intentional.
Stick to the test plan, no additional assertion is needed.

r-recchia commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,434 @@
+        # --------------------------------------------------------------
+
+        # Create Providers list and Add TH4 and TH2 for fabric 1 to Providers list
+        providers_list = [provider_th2_for_fabric1, provider_th4_for_fabric1]
```
Does the order matter? First provider should be TH4

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,434 @@
+
+        # Confirm attribute value did not change
+        asserts.assert_equal(pre_write_val, post_write_val,
+                             "Attribute value changed unexpectedly after failed write")
```
The test plan mentions that Step 5 only involves writing to the `DefaultOTAProviders` attribute on Fabric 1. Fabric 2 is not affected in this step, so reading or verifying Fabric 2 is not required by the test plan. Therefore, no extra validation is needed.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,434 @@
+
+        # Confirm attribute value did not change
+        asserts.assert_equal(pre_write_val, post_write_val,
+                             "Attribute value changed unexpectedly after failed write")
```
The test plan specifies that Step 5 only involves writing and verifying the `DefaultOTAProviders` attribute on Fabric 1. Fabric 2 is not affected in this step, so reading or verifying Fabric 2 here is not required. Therefore, no extra validation is needed for Fabric 2 in Step 5.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,434 @@
+        # - Other fabrics are not affected; their DefaultOTAProviders remain intact.
+        # This explains why we verify Fabric 2 is empty and Fabric 1 still has the original provider.
+
+        # Verify DefaultOTAProviders attribute on the DUT after write (TH4 on Fabric 1) in this case should be the original from TH2 fabric 1
```
According to the test plan, Step 6 only requires verifying that Fabric 2 was updated correctly and that the write succeeded without errors. Validation of Fabric 1 is not required in this step.

juandediosg committed:
Update: Include an assertion in Step 7 to check that UpdatePossible is True.

juandediosg reviewed (commented):
@r-recchia Feedback has been addressed with my clarifications. I will just add one update on step 7 to include an assertion and check that `UpdatePossible` is `True`.

juandediosg reviewed (commented):
Clarified step 5: TH4 is used as the controller performing the write operation, since TH4 is part of the provider list per the test plan.

juandediosg committed:
Refactor based on feedback to align with test plan and verification:
- Use TH as the admin controller across the test
- Refactor Step 5 to match test verification (no TH4 controller)
- Update Step 6 to validate fabric when clearing DefaultOTAProviders
- Update Steps 7 and 8 to read attributes via TH on Fabric 1
- Clarify comments to align with test expectations.

juandediosg reviewed (commented):
Refactor to align test logic with test plan and verification

juandediosg committed:
Update: Copilot comments suggestions.

khodya commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,390 @@
+            post_write_val[0] in [provider_th2_for_fabric1, provider_th4_for_fabric1],
+            "Remaining provider must be either TH2 or TH4 after ConstraintError"
+        )
+
```
read is missing for the second fabric

juandediosg committed:
Update: Step 5 now verifies DefaultOTAProviders on both fabrics after ConstraintError.

juandediosg reviewed (commented):
@khodya  based on your feedback, I added read and validation for `Fabric 2` to ensure the provider remains unchanged after the `ConstraintError` on `Fabric 1`.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,406 @@
+        # Update attribute with empty providers list (TH3 for fabric 2)
+        attr = self.cluster_otar.Attributes.DefaultOTAProviders(value=providers_list_empty)
+
+        # Write updated DefaultOTAProviders attribute to DUT using TH3 (Fabric 2)
```
N/A, Acknowledged current logic is intentional.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,406 @@
+            th_actual_otap_info[0],
+            provider_th4_for_fabric1,
+            "DefaultOTAProviders on Fabric 1 should remain set to TH2 or TH4"
+        )
```
N/A, Acknowledged current logic is intentional based on test plan.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,406 @@
+        self.step(6)
+        # Create Empty Providers list
+        providers_list_empty = []
+        logger.info(f'Step #6 - Providers list updated with provider empty: {providers_list_empty}')
```
N/A, Acknowledged current logic is intentional.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,406 @@
+            setupPinCode=setup_pin_code,
+            filterType=ChipDeviceCtrl.DiscoveryFilterType.LONG_DISCRIMINATOR,
+            filter=long_discriminator)
+        logger.info(f'Step #3 - TH3 Commissioning response: {resp}')
```
Handled by the Framework
N/A, Acknowledged current logic is intentional.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,406 @@
+        th3_actual_otap_info = await self.read_single_attribute_check_success(
+            dev_ctrl=th3,
+            cluster=self.cluster_otar,
+            attribute=self.cluster_otar.Attributes.DefaultOTAProviders)
```
N/A, Acknowledged current logic is intentional.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,406 @@
+            dev_ctrl=th,
+            cluster=self.cluster_otar,
+            attribute=self.cluster_otar.Attributes.DefaultOTAProviders)
+        logger.info(f'Step #5b - AFTER Write - TH DefaultOTAProviders: {post_write_val}')
```
N/A, Acknowledged current logic is intentional.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,406 @@
+            dev_ctrl=th,
+            cluster=self.cluster_otar,
+            attribute=self.cluster_otar.Attributes.DefaultOTAProviders)
+        logger.info(f'Step #5b - BEFORE Write - TH DefaultOTAProviders: {pre_write_val}')
```
N/A, Acknowledged current logic is intentional.

juandediosg commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,406 @@
+        logger.info(f'Step #2 - Read DefaultOTAProviders value after write on DUT (TH): {actual_provider}')
+
+        # Verify the actual provider matches the expected OTA Provider (TH2)
+        # DUT as OTA Requestor (NodeID=2); TH2 as OTA Provider (NodeID=1)
```
N/A, Acknowledged current logic is intentional.

asirko-soft reviewed (approved):
looks good to me

khodya reviewed (approved):
LGTM, probably the last step should be removed from the test plan and test step definitions

cecille commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,406 @@
+        # According to Bug #40294 and Matter specification sections 7.3.3 (Write Path Data Process)
+        # and 10.6.4.3.1 (Lists), when writing a list attribute with multiple providers from the same fabric:
+        # - The write may fail with a ConstraintError if any entry violates constraints.
+        # - Only one valid entry in the list is guaranteed to be preserved ; which entry is preserved
```
@amine - what's the expected behaviour here? Is the list size checked before entries are processed or are entries expected to be processed in order?

The behaviour of a single implementation should be deterministic.

If the ordering of the checks (processing entries vs. checking list size) isn't specified directly, can we MAKE this deterministic by changing the list order?

cecille commented on src/python_testing/TC_SU_4_1.py:
```diff
@@ -0,0 +1,406 @@
+        asserts.assert_equal(
+            th_actual_otap_info[0],
+            provider_th4_for_fabric1,
+            "DefaultOTAProviders on Fabric 1 should remain set to TH2 or TH4"
```
This is checking that it's TH4 only though. The test plan is wrong here - it specifies TH2 or TH4 in step 5, then only TH4 in step 6

r-recchia committed:
Solve comments for TC_SU_4_1

r-recchia committed:
Solve comments for TC_SU_4_1

r-recchia committed:
Revert "Solve comments for TC_SU_4_1"

This reverts commit 8e0d11525b07468684eaaa785cf5e4d671875331.

r-recchia committed:
restore third party

r-recchia committed:
Merge branch 'master' into pr/juandediosg/39980

r-recchia committed:
Fix imports

r-recchia committed:
Merge branch 'master' into TC_SU_4_1_PythonTest

r-recchia committed:
Merge branch 'master' into TC_SU_4_1_PythonTest

r-recchia committed:
Merge branch 'master' into TC_SU_4_1_PythonTest

r-recchia committed:
fix imports

r-recchia committed:
Fix imports

cecille committed:
Merge branch 'master' into TC_SU_4_1_PythonTest

cecille committed:
Merge branch 'master' into TC_SU_4_1_PythonTest

cecille committed:
Merge branch 'master' into TC_SU_4_1_PythonTest
